src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1397,83 @@ struct OptimizeInstructions
     return binary;
   }
 
+  Expression* optimizeBulkMemory(Expression* curr) {
+    if (auto* memCopy = curr->dynCast<MemoryCopy>()) {
+      FeatureSet features = getModule()->features;
+
+      auto destHasSideEffects =
+        EffectAnalyzer(getPassOptions(), features, memCopy->dest)
+          .hasSideEffects();
+
+      // memory.copy(x, x, sz)  ==>  nop
+      if (!destHasSideEffects &&
+          ExpressionAnalyzer::equal(memCopy->dest, memCopy->source)) {
+        return ExpressionManipulator::nop(curr);
+      }
+      // memory.copy(dst, src, C)  ==>  store(dst, load(src))
+      if (auto* csize = memCopy->size->dynCast<Const>()) {
+        auto bytes = csize->value.geti32();
+
+        switch (bytes) {
+          case 0: {
+            if (!destHasSideEffects &&
+                !EffectAnalyzer(getPassOptions(), features, memCopy->source)
+                   .hasSideEffects()) {
+              return ExpressionManipulator::nop(curr);
+            }
+            break;
+          }
+          case 1:
+          case 2:
+          case 4: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
+              memCopy->dest,
+              builder.makeLoad(
+                bytes, false, 0, bytes, memCopy->source, Type::i32),
+              Type::i32);
+          }
+          case 8: {
+            Builder builder(*getModule());
+            return builder.makeStore(
+              bytes, // bytes
+              0,     // offset
+              bytes, // align
+              memCopy->dest,
+              builder.makeLoad(
+                bytes, false, 0, bytes, memCopy->source, Type::i64),
+              Type::i64);
+          }
+          case 16: {
+            if (getPassOptions().shrinkLevel == 0) {
+              // this increase add extra 2 bytes so apply it only for
+              // minimal shrink level
+              if (features.hasSIMD()) {
+                Builder builder(*getModule());
+                return builder.makeStore(
+                  bytes, // bytes
+                  0,     // offset
+                  bytes, // align
+                  memCopy->dest,
+                  builder.makeLoad(
+                    bytes, false, 0, bytes, memCopy->source, Type::v128),
+                  Type::v128);
+              }
+            }
+          }
+          // TODO:
+          // case 3, 5, 6, 7 ... 15? But for -O4?
```

Comment:
I know that SpiderMonkey at least also performs optimizations on constant-length `memory.copy` instructions, so I don't think it will be beneficial to optimize `memory.copy` into loads and stores unless it is very straightforward. Otherwise it would be better to let the engine handle the optimization with its platform-specific knowledge.

Reply:
Good to know. Will remove this TODO comment in that case